### PR TITLE
Expose disk bus

### PR DESCRIPTION
**Pull request by fromanirh**
Make it possible for clients to select the emulated drive exposing the busses supported by libvirt:
ide, scsi/sata, virtio.

Keep the default as virtio, to preserve the current behaviour - bus is related to the device node name, and the old device node name was "vd*", thus virtio.

Pending work: add more tests, and functests for sata. But this require work to make sure the chipset is -or can be set to- Q35.

### CODE REVIEW

File: pkg/api/v1/schema_swagger_generated.go [hunk 1]
```diff
@@ -68,6 +68,7 @@ func (DiskTarget) SwaggerDoc() map[string]string {
 	return map[string]string{
 		"dev":      "Device indicates the \"logical\" device name. The actual device name\nspecified is not guaranteed to map to the device name in the guest OS. Treat\nit as a device ordering hint.",
 		"readonly": "ReadOnly\nDefaults to false",
+		"bus":      "Bus indicates specifies the type of disk device to emulate.",
```
Comment (stu-gott) [inline pkg/api/v1/schema_swagger_generated.go]: nit: Remove one of these two words: "indicates" or "specifies".

File: pkg/virt-handler/virtwrap/api/converter_test.go [hunk 2]
```diff
@@ -399,6 +402,8 @@ var _ = Describe("Converter", func() {
 
 		It("should be converted to a libvirt Domain with vm defaults set", func() {
 			v1.SetObjectDefaults_VirtualMachine(vm)
+			log.Printf(vmToDomainXML(vm, c))
```
[the remark this answers is not in the document]
Comment (fromanirh) [reply pkg/virt-handler/virtwrap/api/converter_test.go]: Sure, sneaked in by mistake

File: pkg/api/v1/schema_swagger_generated.go (hunk 1 above)
Comment (fromanirh) [reply pkg/api/v1/schema_swagger_generated.go]: Sure

File: api/openapi-spec/swagger.json [hunk 3]
```diff
@@ -1925,7 +1925,14 @@
     }
    },
    "v1.DiskTarget": {
+    "required": [
+     "bus"
+    ],
     "properties": {
+     "bus": {
+      "description": "Bus indicates specifies the type of disk device to emulate.",
```
[the remark this answers is not in the document]
Comment (fromanirh) [reply api/openapi-spec/swagger.json]: Sure, will remove

File: pkg/api/v1/defaults.go [hunk 4]
```diff
@@ -82,6 +82,24 @@ func SetDefaults_FloppyTarget(obj *FloppyTarget) {
 	}
 }
 
+func SetDefaults_DiskTarget(obj *DiskTarget) {
+	if obj.Bus == "" {
+		obj.Bus = "virtio"
```
[the remark this answers is not in the document]
Comment (fromanirh) [reply pkg/api/v1/defaults.go]: We totally should, but before that we need either
- to be able to select q35, 'sata' won't work on i44fx AFAIK
- to default to q35 implicitely

File: pkg/api/v1/schema.go [hunk 5]
```diff
@@ -120,6 +120,8 @@ type DiskTarget struct {
 	// ReadOnly
 	// Defaults to false
 	ReadOnly bool `json:"readonly,omitempty"`
+	// Bus indicates specifies the type of disk device to emulate.
+	Bus string `json:"bus"`
```
Comment (lukas-bednar) [inline pkg/api/v1/schema.go]: If `Bus` field has a default option, then I would add `omitempty` option.

File: api/openapi-spec/swagger.json [hunk 6]
```diff
@@ -1925,7 +1925,14 @@
     }
    },
    "v1.DiskTarget": {
+    "required": [
+     "bus"
```
Comment (davidvossel) [inline api/openapi-spec/swagger.json]: This shouldn't be a required argument.

Comment (fromanirh) [reply api/openapi-spec/swagger.json]: We can infer the "bus" value from the VirtualMachine Spec, nut we need one set of attributes to distinguish what the VM is gonna used for.
The simplest example is the guest OS. Assuming we use a Q35 chipset (as we should):
- windows guest OS: -> bus = sata
- Linux (any) guest OS: -> bus = virtio

So the question becomes: what we can reliably use as set of attributes?

**Comment by fromanirh** [comment]
@davidvossel , maybe we can piggyback on the presence of hyperv feature.
HyperV feature enabled -> tune for windows -> in this example, bus=sata
HyperV feature disabled -> tune for linux -> in this example, bus=virtio

The only drawback I can see, is nothing prevents anyone to run linux on hyperv-enabled VMs. But at libvirt level and beyond, the extra settings should all go ignored, so the VM will still run, albeit unoptimized. Could be good enough for now?

**Comment by fabiand** [comment]
First, I agree that the bus attribute should not be mandatory. I missed that.

However, one reason for KubeVirt is a migration story - to get existing VMs from somewhere to Kubernetes to run them alongside pods.
In order to achieve this, KubeVirt' strong point is to allow creating VMs as close as possible as they were on the former mgmt platform.
In order to achieve this, a user, or a migration system, needs to have the full control over the defails. I.e. this bus type.

And in general to me it does not look like the bus type can be generally inferred, as it also heavily depends on the guest support.
I.e.

- Inferring sata for "hyperv" VMs works, but you will not be able to use virti once the VM has virtio drivers
- Inferring virtion for linux VMs (let's ignore how we reliably infer linux vms from the guest ABI) might work, but good old RHEL 4 might not have virtio drivers, thus would be unbootable

Bottom line to me: The API needs to permit to set the bus. But we should be able to do the work for our users.
I.e. with presets, libosinfo, and maybe some more libvirtd help we can come up with a way to make it really slick to let the system set all the defaults.

**Comment by rmohr** [comment]
Let's make sure that we ask ourselves always a few questions before we add something like this. In this case, I especially wonder if we ever have the situation (legacy vm or not), where one mixes the bus types for disks. With other words, if that use-case even exists. It might be enough to just specify the bus once for all disks. @michalskrivanek @fromanirh thoughts?

**Comment by fromanirh** [comment]
@rmohr I don't have clear examples at hand on which one may or may not want all the drives attached to a VM to have the same bus.

But the way the device is exposed to the guest (determined by the bus type) changes the feature set of the said device, so it is not "just" a performance thing: one may want to have one scsi disk to use the scsi passthrough, while other drives are ok with virtio.

In general, considering the upgrade path @fabiand mentioned and that we target the VM use case, I think that restricting all the drives to have the same bus type seems a bit too artificial restriction *and* I don't think that would simplify our flows significantly.

File: pkg/api/v1/defaults.go (hunk 4 above)
[the remark this answers is not in the document]
Comment (rmohr) [reply pkg/api/v1/defaults.go]: so we should use "ide" for default if i44fx is selected and "sata" if q35?

[the remark this answers is not in the document]
Comment (fromanirh) [reply pkg/api/v1/defaults.go]: this is indeed the most correct fix. But AFAIR we want to always use q35, don't we?

File: api/openapi-spec/swagger.json (hunk 6 above)
Comment (rmohr) [reply api/openapi-spec/swagger.json]: For as long as we don't have the presets in place, let's default to the more compatible buses. That means imo sata if 135 if selected and probably ide if it is an older machine type.

File: pkg/api/v1/defaults.go [hunk 7]
```diff
@@ -82,6 +82,24 @@ func SetDefaults_FloppyTarget(obj *FloppyTarget) {
...
+	}
+	if obj.Device == "" {
+		switch obj.Bus {
+		case "ide":
+			obj.Device = "hda"
```
Comment (rmohr) [inline pkg/api/v1/defaults.go]: That does not look right.  That would give every disk the same device name hint.

File: pkg/api/v1/defaults.go (hunk 4 above)
[the remark this answers is not in the document]
Comment (rmohr) [reply pkg/api/v1/defaults.go]: always as default, it can be overwritten anytime.

File: pkg/api/v1/schema.go (hunk 5 above)
Comment (fromanirh) [reply pkg/api/v1/schema.go]: Done

File: api/openapi-spec/swagger.json (hunk 6 above)
Comment (fromanirh) [reply api/openapi-spec/swagger.json]: Done

File: pkg/virt-handler/virtwrap/api/converter_test.go [hunk 8]
```diff
@@ -330,7 +330,7 @@ var _ = Describe("Converter", func() {
       <source protocol="iscsi" name="iqn.2013-07.com.example:iscsi-nopool/2">
         <host name="example.com" port="3260"></host>
       </source>
-      <target bus="virtio" dev="vda"></target>
+      <target bus="ide" dev="hdg"></target>
```
Comment (fromanirh) [inline pkg/virt-handler/virtwrap/api/converter_test.go]: I want to add more tests in a future upload - or even in a future PR if the reviewers prefer so.

File: pkg/virt-handler/virtwrap/api/converter.go [hunk 9]
```diff
@@ -48,6 +49,51 @@ func Convert_v1_Disk_To_api_Disk(diskDevice *v1.Disk, disk *Disk) error {
...
+	switch bus {
+	case "virtio":
+		prefix = "vd"
+	case "sata", "scsi":
+		prefix = "sd"
+	case "ide":
+		prefix = "hd"
+	default:
+		// we should never get there
+		return ""
```
Comment (stu-gott) [inline pkg/virt-handler/virtwrap/api/converter.go]: I feel that this should be an error condition. If this code block is unexpected then failing silently is not the ideal choice--we'd end up with a disk named `a`... :)

If there's at least a reasonable default (for example "hd"?) this should still be logged somewhere.

File: pkg/virt-handler/virtwrap/api/converter.go [hunk 10]
```diff
@@ -48,6 +49,51 @@ func Convert_v1_Disk_To_api_Disk(diskDevice *v1.Disk, disk *Disk) error {
...
+	}
+	return formatDeviceName(prefix, index)
+}
+
+// port of http://elixir.free-electrons.com/linux/v4.15/source/drivers/scsi/sd.c#L3211
+func formatDeviceName(prefix string, index int) string {
+	base := int('z' - 'a' + 1)
+	name := prefix
+
+	for index >= 0 {
```
Comment (stu-gott) [inline pkg/virt-handler/virtwrap/api/converter.go]: This is incorrect. The letters in the output name are backwards.

the 27th disk should be `sdab` not `sdba`.
the 28th disk should be `sdac` not `sdca`...

In the algorithm referenced in the comment, there was a memcpy at the end to reverse the output.

File: pkg/virt-handler/virtwrap/api/converter.go [hunk 11]
```diff
@@ -48,6 +49,51 @@ func Convert_v1_Disk_To_api_Disk(diskDevice *v1.Disk, disk *Disk) error {
 	return nil
 }
 
+func diskBusFromMachine(bus, machine string) string {
+	if bus != "" {
+		// use what it was explicitely selected
```
Comment (stu-gott) [inline pkg/virt-handler/virtwrap/api/converter.go]: Nit: "use what was explicitly selected"

File: pkg/virt-handler/virtwrap/api/converter.go [hunk 12]
```diff
@@ -287,9 +333,32 @@ func Convert_v1_VirtualMachine_To_api_Domain(vm *v1.VirtualMachine, domain *Doma
 		volumes[volume.Name] = volume.DeepCopy()
 	}
 
-	for _, disk := range vm.Spec.Domain.Devices.Disks {
+	machine := "pc" // QEMU default
+	if vm.Spec.Domain.Machine != nil {
```
Comment (stu-gott) [inline pkg/virt-handler/virtwrap/api/converter.go]: If `Spec.Domain.Machine` is `nil` would it make sense to initialize it?

File: pkg/virt-handler/virtwrap/api/converter.go (hunk 10 above)
Comment (fromanirh) [reply pkg/virt-handler/virtwrap/api/converter.go]: Right, good catch! fixed

File: pkg/virt-handler/virtwrap/api/converter.go (hunk 12 above)
Comment (fromanirh) [reply pkg/virt-handler/virtwrap/api/converter.go]: we handle Spec.Domain.Machine properly in line 402. I added this (minor) duplication to keep this patch minimal, I'd like some refactoring later (next upload or future PR as reviewers prefer)

File: pkg/virt-handler/virtwrap/api/converter.go [hunk 13]
```diff
@@ -20,11 +20,12 @@ type ConverterContext struct {
...
 }
 
-func Convert_v1_Disk_To_api_Disk(diskDevice *v1.Disk, disk *Disk) error {
+func Convert_v1_Disk_To_api_Disk(diskDevice *v1.Disk, disk *Disk, machine string, index int) error {
 
 	if diskDevice.Disk != nil {
 		disk.Device = "disk"
-		disk.Target.Device = diskDevice.Disk.Device
+		disk.Target.Bus = diskBusFromMachine(diskDevice.Disk.Bus, machine)
+		disk.Target.Device = diskDeviceFromBusAndIndex(diskDevice.Disk.Device, disk.Target.Bus, index)
```
Comment (rmohr) [inline pkg/virt-handler/virtwrap/api/converter.go]: I know, doing this in converter.go was my suggestion, but I think we need to do this on the cluster level. Also that q35 should be set as default on the cluster level. I know that inter-dependencies between e.g. machine type and disk are not so nice to express in the cluster-level defaulter, but you can squeeze it probably into `SetDefault_VirtualMachine`.

File: pkg/virt-handler/virtwrap/api/converter.go [hunk 14]
```diff
@@ -48,6 +49,51 @@ func Convert_v1_Disk_To_api_Disk(diskDevice *v1.Disk, disk *Disk) error {
...
+	// see /path/to/qemu-kvm -machine help
+	if strings.HasPrefix(machine, "pc-q35") || strings.HasPrefix(machine, "q35") {
+		return "sata"
+	}
+	// safe fallback for x86_64, but very slow
+	return "ide"
+}
+
+func diskDeviceFromBusAndIndex(deviceName, bus string, index int) string {
+	if deviceName != "" {
```
Comment (rmohr) [inline pkg/virt-handler/virtwrap/api/converter.go]: To make this really right, we would need do consider, that a person might set the deviceName on e.g. the third disk to "sda", which means we would then have "sda" on the first and the third disk.

Comment (fromanirh) [reply pkg/virt-handler/virtwrap/api/converter.go]: Yes, we would need to scan all the drives and to somehow mark device names already taken. It is not that hard, just another look + map, but I wonder if we need this complexity at all. I mean, *here* I'm having one hard time finding for one use case. Thoughts?

File: pkg/virt-handler/virtwrap/api/converter.go (hunk 13 above)
Comment (fromanirh) [reply pkg/virt-handler/virtwrap/api/converter.go]: I agree with the concept, but does this default fixing belong here or is better be served by another PR?

Comment (rmohr) [reply pkg/virt-handler/virtwrap/api/converter.go]: I think it makes sense to properly implement it right right now from the defaulting standpoint. No need for extra PRs.

File: pkg/virt-handler/virtwrap/api/converter.go (hunk 14 above)
Comment (rmohr) [reply pkg/virt-handler/virtwrap/api/converter.go]: You can also completely remove the device name from the cluster api and just do the deviceName logic on the node where you are doing it right now. @fabiand @davidvossel thoughts?

File: pkg/virt-handler/virtwrap/api/converter.go (hunk 13 above)
Comment (fabiand) [reply pkg/virt-handler/virtwrap/api/converter.go]: Hm. If we do this on the cluster level, then it will end up in any object after creation.

And afaik we are not doing this with anything else yet, are we?

File: pkg/virt-handler/virtwrap/api/converter.go (hunk 14 above)
Comment (fabiand) [reply pkg/virt-handler/virtwrap/api/converter.go]: I'm in favor of removing the whole deviceName thing from the public API.

To me it leads more to confusion (as it's depending on guest support) than helping.

Comment (fromanirh) [reply pkg/virt-handler/virtwrap/api/converter.go]: great, I'll make the code expose bus (which is totally fair and helpful IMO) instead.

**Comment by rmohr** [comment]
So to sum all discussions up:

Bus is exposed for disks on the cluster level (including defaulting based on machine type) and devName will be removed from the cluster API and will be set based on the array index on the domain only (node level).

**Comment by fromanirh** [comment]
Great, implementing the changes

File: pkg/api/v1/defaults.go [hunk 15]
```diff
@@ -117,6 +119,46 @@ func SetDefaults_VirtualMachine(obj *VirtualMachine) {
 	if obj.Spec.Domain.Features == nil {
 		obj.Spec.Domain.Features = &Features{}
 	}
+	// TODO: not fond of this code, suggestions welcome
+	if obj.Spec.Domain.Machine == nil {
+		obj.Spec.Domain.Machine = &Machine{}
```
Comment (rmohr) [inline pkg/api/v1/defaults.go]: If we set a default always anyway, you can mache Machine a non-pointer and just check if Machine.Type is an empty string.

File: pkg/api/v1/schema_test.go [hunk 16]
```diff
@@ -114,14 +117,14 @@ var exampleJSON = `{
             "name": "disk0",
             "volumeName": "volume0",
             "disk": {
-              "dev": "vda"
+              "bus": "virtio"
```
Comment (rmohr) [inline pkg/api/v1/schema_test.go]: Are you sure that virtio was the default before?

**Comment by rmohr** [comment]
ok to test

File: pkg/api/v1/defaults.go (hunk 15 above)
Comment (fromanirh) [reply pkg/api/v1/defaults.go]: done

File: pkg/api/v1/schema_test.go (hunk 16 above)
Comment (fromanirh) [reply pkg/api/v1/schema_test.go]: yes, albeit implicitely since we were expecting "vd*" device name.

File: pkg/api/v1/schema.go [hunk 17]
```diff
@@ -121,30 +121,24 @@ type DiskDevice struct {
...
 	// Defaults to false
 	ReadOnly bool `json:"readonly,omitempty"`
 }
 
 type LunTarget struct {
-	// Device indicates the "logical" device name. The actual device name
-	// specified is not guaranteed to map to the device name in the guest OS. Treat
-	// it as a device ordering hint.
-	Device string `json:"dev,omitempty"`
+	// Bus indicates the type of disk device to emulate.
```
Comment (rmohr) [inline pkg/api/v1/schema.go]: Please enumerate all the bus types we support. Right now we don't have validation, but we will have it as soon as we have the validation webhook in place and we should tell the user what we will support. The comment here directly goes into the api reference. Does this one solely support scsi or also virtio?

File: pkg/api/v1/schema.go [hunk 18]
```diff
@@ -121,30 +121,24 @@ type DiskDevice struct {
...
 	ReadOnly bool `json:"readonly,omitempty"`
 }
 
 type FloppyTarget struct {
-	// Device indicates the "logical" device name. The actual device name
-	// specified is not guaranteed to map to the device name in the guest OS. Treat
-	// it as a device ordering hint.
-	Device string `json:"dev,omitempty"`
+	// Bus indicates the type of disk device to emulate.
+	Bus string `json:"bus,omitempty"`
```
Comment (rmohr) [inline pkg/api/v1/schema.go]: You wrote further down below that "fd" is our only choice. Does it then really make sense to have the bus exposed on the floppy drive?

File: pkg/api/v1/defaults.go [hunk 19]
```diff
@@ -117,6 +119,42 @@ func SetDefaults_VirtualMachine(obj *VirtualMachine) {
...
+		}
+		if disk.CDRom != nil && disk.CDRom.Bus == "" {
+			disk.CDRom.Bus = bus
+		}
+		if disk.LUN != nil && disk.LUN.Bus == "" {
+			disk.LUN.Bus = bus
+		}
+		if disk.Floppy != nil && disk.Floppy.Bus == "" {
+			// no real choice here
+			disk.Floppy.Bus = "fd"
```
Comment (rmohr) [inline pkg/api/v1/defaults.go]: Is that the only reasonable value for floppy in general?

File: tests/vm_configuration_test.go [hunk 20]
```diff
@@ -75,4 +75,75 @@ var _ = Describe("Configurations", func() {
 			Expect(err).ToNot(HaveOccurred())
 		}, 300)
 	})
+
+	Context("New VM with explicitly set VirtIO drives", func() {
```
Comment (rmohr) [inline tests/vm_configuration_test.go]: Could you attach a few more disks to the one vm with different bus types and check the result? One or two tests which check multiple buses should be good enough (good for the overall test time since we don't have to boot the vm that often).


Something like:

```yaml
apiVersion: kubevirt.io/v1alpha1
kind: VirtualMachine
metadata:
  name: testvm-ephemeral
spec:
  domain:
    resources:
      requests:
        memory: 64M
    devices:
      disks:
      - name: disk1
        volumeName: disk1
        disk:
          bus: ide
      - name: disk2
        volumeName: disk2
        disk:
          bus: sata
      - name: disk3
        volumeName: disk3
        cdrom:
          bus: virtio
      - name: disk4
        volumeName: disk4
        floppy: {}
      - name: disk5
        volumeName: disk5
        disk:
          bus: virtio
  volumes:
    - name: disk1
      registryDisk:
        image: kubevirt/cirros-registry-disk-demo:devel
    - name: disk2
      registryDisk:
        image: kubevirt/cirros-registry-disk-demo:devel
    - name: disk3
      registryDisk:
        image: kubevirt/cirros-registry-disk-demo:devel
    - name: disk4
      registryDisk:
        image: kubevirt/cirros-registry-disk-demo:devel
    - name: disk5
      registryDisk:
        image: kubevirt/cirros-registry-disk-demo:devel
```

File: pkg/api/v1/schema.go (hunk 17 above)
Comment (fromanirh) [reply pkg/api/v1/schema.go]: OK, will copy the content of https://github.com/libvirt/libvirt/blob/master/docs/schemas/domaincommon.rng#L1785 (libvirt schema) AND link it as reference.

File: pkg/api/v1/schema.go (hunk 18 above)
Comment (fromanirh) [reply pkg/api/v1/schema.go]: probably not for the time being

File: tests/vm_configuration_test.go (hunk 20 above)
Comment (fromanirh) [reply tests/vm_configuration_test.go]: Of course, adding such test.

File: pkg/api/v1/defaults.go (hunk 19 above)
Comment (fromanirh) [reply pkg/api/v1/defaults.go]: I'm not 100% sure that it is the only reasonable value for floppies, e.g we can attach one floppy through usb. But surely it is a supported value, and, looking at libvirt sources, seems the preferred one. And it should be "fdc" (see below)

File: pkg/api/v1/schema.go (hunk 17 above)
Comment (rmohr) [reply pkg/api/v1/schema.go]: Let's start with adding those which make sense for us right now. The link just provides what is accepted in general, independent if we are talking about disks, luns, cdrom and floppy. I think the content of the link is just confusing. Directly stating what makes sense is more useful.

File: pkg/api/v1/schema.go [hunk 21]
```diff
@@ -161,6 +167,9 @@ const (
 
 type CDRomTarget struct {
 	// Bus indicates the type of disk device to emulate.
+	// supported values: virtio, sata, scsi, ide
+	// see libvirt schema here:
+	// https://github.com/libvirt/libvirt/blob/v3.7-maint/docs/schemas/domaincommon.rng#L1693
```
Comment (rmohr) [inline pkg/api/v1/schema.go]: I thin the schema is really not helpful.

Comment (fromanirh) [reply pkg/api/v1/schema.go]: fair enough, I will remove the references.

**Comment by rmohr** [comment]
@fromanirh there were some issues in the defaulting code and the test-vm setup. I am not yet completely happy with the test (since we can't detect all devices), but everything seems to work now.

**Comment by fromanirh** [comment]
@rmohr uhm, I managed to play a bit with the code - applying your changes and other locally and still not working properly for me. How do you want to proceed with this PR? I mean, what's the next step?

**Comment by rmohr** [comment]
@fromanirh what is not working?

**Comment by fromanirh** [comment]
@rmohr false alarm, sorry. The failure I got looks unrelated:

Vmlifecycle
/root/go/src/kubevirt.io/kubevirt/tests/vmlifecycle_test.go:41
  New VM given
  /root/go/src/kubevirt.io/kubevirt/tests/vmlifecycle_test.go:60
    Should start the VM on POST
    /root/go/src/kubevirt.io/kubevirt/tests/vmlifecycle_test.go:87
------------------------------
• [SLOW TEST:13.020 seconds]
Vmlifecycle
/root/go/src/kubevirt.io/kubevirt/tests/vmlifecycle_test.go:41
  New VM given
  /root/go/src/kubevirt.io/kubevirt/tests/vmlifecycle_test.go:60
    Virt-launcher should attach to a started VM
    /root/go/src/kubevirt.io/kubevirt/tests/vmlifecycle_test.go:95
------------------------------
level=info timestamp=2018-02-08T12:23:49.552690Z pos=utils.go:172 component=tests msg="VM defined."
level=info timestamp=2018-02-08T12:23:49.561787Z pos=utils.go:172 component=tests msg="VM started."
level=info timestamp=2018-02-08T12:23:49.595479Z pos=utils.go:183 component=tests msg="VM defined."
level=info timestamp=2018-02-08T12:23:49.595575Z pos=utils.go:183 component=tests msg="VM started."
level=info timestamp=2018-02-08T12:23:49.623489Z pos=utils.go:183 component=tests msg="VM defined."
level=info timestamp=2018-02-08T12:23:49.670067Z pos=utils.go:183 component=tests msg="Signaled Graceful Shutdown"
level=info timestamp=2018-02-08T12:23:49.686470Z pos=utils.go:183 component=tests msg="Signaled Graceful Shutdown"
level=info timestamp=2018-02-08T12:23:49.698524Z pos=utils.go:183 component=tests msg="Signaled Graceful Shutdown"
level=error timestamp=2018-02-08T12:23:54.923004Z pos=utils.go:181 component=tests reason="Unexpected Warning event recieved." msg="server error. command Launcher.Kill failed: virError(Code=42, Domain=10, Message='Domain not found: no domain with matching uuid 'f015b329-7a83-4be9-955e-f1c57109d42d' (kubevirt-test-default_testvmf8cml)')"
•... Timeout [45.121 seconds]
Vmlifecycle
/root/go/src/kubevirt.io/kubevirt/tests/vmlifecycle_test.go:41
  New VM given
  /root/go/src/kubevirt.io/kubevirt/tests/vmlifecycle_test.go:60
    if grace period greater than 0, graceful shutdown should be attempted on delete [It]
    /root/go/src/kubevirt.io/kubevirt/tests/vmlifecycle_test.go:108

    Timed out

    /root/go/src/kubevirt.io/kubevirt/tests/vmlifecycle_test.go:108
------------------------------
• [SLOW TEST:13.450 seconds]
Vmlifecycle
/root/go/src/kubevirt.io/kubevirt/tests/vmlifecycle_test.go:41
  New VM given
  /root/go/src/kubevirt.io/kubevirt/tests/vmlifecycle_test.go:60
    New VM which can't be started
    /root/go/src/kubevirt.io/kubevirt/tests/vmlifecycle_test.go:154
      Should retry starting the VM
      /root/go/src/kubevirt.io/kubevirt/tests/vmlifecycle_test.go:156
------------------------------
• [SLOW TEST:14.355 seconds]

**Comment by rmohr** [comment]
retest this please

**Comment by rmohr** [comment]
@fabiand I just saw that one can still set the bus on a floppy. could you either remove the bus field there, or document which values can be used?

**Comment by fabiand** [comment]
@rmohr I suppose @fromanirh ?

**Comment by fromanirh** [comment]
@rmohr @fabiand  sure will fix asap
